File: examples/example-py/bare.py
import asyncio
import json
import os
from dataclasses import dataclass

import httpx
from nearai_inference_sdk import (
    AttestationClient,
    AttestationVerifiers,
    E2eeModelKey,
    ImageProvenancePolicy,
    MeasuredDeployment,
    VerifiedGatewayAttestation,
    VerifiedModelAttestation,
    create_pinned_tls_client,
    prepare_e2ee_chat_request,
    verify_deployment_image_provenance,
    verify_gateway_attestation,
    verify_gateway_response,
    verify_model_attestation,
    verify_model_response,
)


BASE_URL = 'https://cloud-api.near.ai/v1/'
MODEL = 'z-ai/glm-5.3-flash'
SIGNING_ALGO = 'ed25519'
GATEWAY_IMAGE_POLICIES = {
    'nearaidev/cloud-api': ImageProvenancePolicy(
        repository='nearai/cloud-api', workflow='.github/workflows/build.yml'
    ),
    'nearaidev/cvm-ingress': ImageProvenancePolicy(
        repository='nearai/cvm-ingress', workflow='.github/workflows/build-push.yml'
    ),
    'nearaidev/dstack-vpc': ImageProvenancePolicy(
        repository='nearai/dstack-vpc', workflow='.github/workflows/build.yml'
    ),
    'nearaidev/dstack-vpc-client': ImageProvenancePolicy(
        repository='nearai/dstack-vpc-client', workflow='.github/workflows/build.yml'
    ),
}


@dataclass
class Completion:
    id: str
    request_body: bytes
    response_body: bytes


@dataclass(frozen=True)
class E2eeRecipient:
    attestation: VerifiedModelAttestation
    key: E2eeModelKey


class RecordingStream(httpx.AsyncByteStream):
    """Keep ciphertext while the E2EE decryptor consumes the response stream."""

    def __init__(self, source: httpx.AsyncByteStream) -> None:
        self.source = source
        self.chunks: list[bytes] = []

    async def __aiter__(self):
        async for chunk in self.source:
            self.chunks.append(chunk)
            yield chunk

    async def aclose(self) -> None:
        await self.source.aclose()


async def verify_gateway_images(deployment: MeasuredDeployment) -> None:
    await verify_deployment_image_provenance(
        deployment.app_compose, GATEWAY_IMAGE_POLICIES
    )


async def fetch_and_verify_gateway(
    client: AttestationClient,
) -> VerifiedGatewayAttestation:
    # Fetching checks the API response and records the client nonce/TLS peer;
    # verification authenticates the quote, binding, and measured configuration.
    fetched = await client.fetch_gateway_attestation(signing_algo=SIGNING_ALGO)
    verified = await verify_gateway_attestation(
        fetched.attestation,
        fetched.client_binding,
        verifiers=AttestationVerifiers(deployment=verify_gateway_images),
    )
    print('Gateway deployment and image provenance: verified.')
    return verified


async def fetch_and_verify_models(
    client: AttestationClient,
) -> list[VerifiedModelAttestation]:
    fetched = await client.fetch_model_attestations(MODEL, signing_algo=SIGNING_ALGO)
    if not fetched.attestations:
        raise RuntimeError('Gateway returned no model attestations')

    # These are model reports fetched through the Gateway, not direct model TLS
    # connections. Verify every report before choosing an encryption recipient.
    models = []
    for attestation in fetched.attestations:
        verified = await verify_model_attestation(attestation, fetched.client_binding)
        models.append(verified)
    print(f'Model deployments: verified {len(models)}.')
    return models


def find_e2ee_recipient(
    models: list[VerifiedModelAttestation],
) -> E2eeRecipient:
    # No response signature exists yet; choose a verified key for the request.
    for model in models:
        if (
            model.signer.signing_algo == SIGNING_ALGO
            and model.signing_public_key is not None
        ):
            return E2eeRecipient(
                attestation=model,
                key=E2eeModelKey(
                    signing_algo=SIGNING_ALGO, public_key=model.signing_public_key
                ),
            )
    raise RuntimeError('No verified model public key is available for E2EE')


async def send_completion(
    client: httpx.AsyncClient, api_key: str, recipient: E2eeRecipient, *, stream: bool
) -> Completion:
    request = httpx.Request(
        'POST',
        BASE_URL + 'chat/completions',
        headers={'Authorization': f'Bearer {api_key}', 'Accept-Encoding': 'identity'},
        json={
            'model': MODEL,
            'messages': [{'role': 'user', 'content': 'Reply with the word ok.'}],
            'max_completion_tokens': 8,
            'stream': stream,
        },
    )
    # The helper encrypts supported fields and creates a fresh response key. It
    # does not send the request or repeat attestation verification.
    prepared = await prepare_e2ee_chat_request(request, recipient.key)
    request_body = prepared.request.content
    response = await client.send(prepared.request, stream=True)
    response.raise_for_status()
    recording = RecordingStream(response.stream)
    response.stream = recording
    decrypted = await prepared.decrypt_response(response)
    completion_id = None
    stream_complete = False
    try:
        if stream:
            async for line in decrypted.aiter_lines():
                if not line.startswith('data:'):
                    continue
                data = line.removeprefix('data:').strip()
                if data == '[DONE]':
                    stream_complete = True
                    continue
                chunk = json.loads(data)
                completion_id = chunk.get('id', completion_id)
                for choice in chunk.get('choices', []):
                    print(
                        choice.get('delta', {}).get('content') or '', end='', flush=True
                    )
            print()
        else:
            await decrypted.aread()
            body = decrypted.json()
            completion_id = body['id']
            print(body['choices'][0]['message'].get('content') or '')
    finally:
        await decrypted.aclose()
        await response.aclose()

    if completion_id is None:
        raise RuntimeError('Completion response did not contain an ID')
    if stream and not stream_complete:
        raise RuntimeError('Streaming completion ended before [DONE]')
    # Signatures cover the encrypted HTTP bytes, not the displayed plaintext.
    return Completion(completion_id, request_body, b''.join(recording.chunks))


async def run_non_streaming_example(client, api_key, recipient) -> Completion:
    return await send_completion(client, api_key, recipient, stream=False)


async def run_streaming_example(client, api_key, recipient) -> Completion:
    return await send_completion(client, api_key, recipient, stream=True)


async def verify_completion_receipt(
    client: AttestationClient,
    completion: Completion,
    gateway: VerifiedGatewayAttestation,
    recipient: E2eeRecipient,
) -> None:
    signature = await client.fetch_completion_signature(
        completion.id, signing_algo=SIGNING_ALGO
    )
    if signature.kind == 'provider_tee':
        verify_model_response(
            completion.request_body,
            completion.response_body,
            signature,
            recipient.attestation,
        )
    else:
        # A Gateway signature proves these bytes were signed by its verified key;
        # it does not prove which model deployment generated the content.
        verify_gateway_response(
            completion.request_body, completion.response_body, signature, gateway
        )
    print(f'Verified {signature.kind} response.')


async def main() -> None:
    api_key = os.environ['NEARAI_API_KEY']
    attestation_client = AttestationClient(api_key, base_url=BASE_URL)
    gateway = await fetch_and_verify_gateway(attestation_client)
    models = await fetch_and_verify_models(attestation_client)
    recipient = find_e2ee_recipient(models)

    # Native clients should pin Chat TLS to the authenticated Gateway SPKI.
    # Browser clients cannot inspect the peer certificate and skip this step.
    async with create_pinned_tls_client(
        gateway.tls_binding.spki_fingerprint
    ) as pinned_tls_client:
        completion = await run_non_streaming_example(
            pinned_tls_client, api_key, recipient
        )
        await verify_completion_receipt(
            attestation_client, completion, gateway, recipient
        )

        completion = await run_streaming_example(
            pinned_tls_client, api_key, recipient
        )
        await verify_completion_receipt(
            attestation_client, completion, gateway, recipient
        )


if __name__ == '__main__':
    asyncio.run(main())

File: examples/example-js/gateway/bare.ts
import {
  AttestationClient,
  createPinnedTlsFetch,
  prepareE2eeChatRequest,
  verifyDeploymentImageProvenance,
  verifyGatewayAttestation,
  verifyGatewayResponse,
  verifyModelAttestation,
  verifyModelResponse,
} from '@nearai/inference-sdk/node';
import type {
  E2eeModelKey,
  GatewayTlsBinding,
  ImageProvenancePolicy,
  VerifiedGatewayAttestation,
  VerifiedModelAttestation,
} from '@nearai/inference-sdk/node';

// This example composes the SDK's standalone APIs: verify deployments, encrypt
// and send Chat requests, decrypt responses, then verify their signatures.
const BASE_URL = 'https://cloud-api.near.ai/v1/';
const MODEL = 'z-ai/glm-5.3-flash';
const SIGNING_ALGO = 'ed25519';

// These policies require GitHub build provenance for four Gateway images,
// not model runtime images. Add a reviewed `commit` to pin a specific build.
const GATEWAY_IMAGE_POLICIES: Record<string, ImageProvenancePolicy> = {
  'nearaidev/cloud-api': {
    repository: 'nearai/cloud-api',
    workflow: '.github/workflows/build.yml',
  },
  'nearaidev/cvm-ingress': {
    repository: 'nearai/cvm-ingress',
    workflow: '.github/workflows/build-push.yml',
  },
  'nearaidev/dstack-vpc': {
    repository: 'nearai/dstack-vpc',
    workflow: '.github/workflows/build.yml',
  },
  'nearaidev/dstack-vpc-client': {
    repository: 'nearai/dstack-vpc-client',
    workflow: '.github/workflows/build.yml',
  },
};

await main();

async function main(): Promise<void> {
  const apiKey = process.env.NEARAI_API_KEY;
  if (!apiKey) {
    throw new Error('NEARAI_API_KEY is required');
  }

  const client = new AttestationClient({ apiKey, baseUrl: BASE_URL });

  // Verify the Gateway and every returned model report before sending a prompt.
  // A failed check throws and stops this flow before either Chat request.
  const gateway = await fetchAndVerifyGateway(client);
  const models = await fetchAndVerifyModelAttestations(client);

  // Encryption needs only the verified public key and its algorithm. Keep the
  // full model verification result separately for the response signature check.
  const model = findModelAttestationForEncryption(models);
  const modelKey: E2eeModelKey = {
    signingAlgo: model.signer.signingAlgo,
    publicKey: model.signingPublicKey,
  };

  // TLS pinning is recommended in Node. In browsers, use regular fetch instead;
  // browser APIs do not expose the TLS peer certificate needed for this check.
  const pinnedTlsFetch = createGatewayPinnedTlsFetch(gateway.tlsBinding);

  // Non-streaming: decrypt and display the JSON response, then verify its signature.
  // To display only verified output, defer rendering until verification succeeds.
  const nonStreamingCompletion = await runNonStreamingExample({
    apiKey,
    pinnedTlsFetch,
    modelKey,
  });
  await verifyCompletionResponse({
    client,
    completion: nonStreamingCompletion,
    gateway,
    model,
  });

  // Streaming: display decrypted SSE as it arrives. Signature verification needs
  // the complete response bytes, so it runs after the stream finishes.
  const streamingCompletion = await runStreamingExample({
    apiKey,
    pinnedTlsFetch,
    modelKey,
  });
  await verifyCompletionResponse({
    client,
    completion: streamingCompletion,
    gateway,
    model,
  });
}

async function runNonStreamingExample({
  apiKey,
  pinnedTlsFetch,
  modelKey,
}: CompletionExampleParams): Promise<Completion> {
  return sendCompletion({
    apiKey,
    stream: false,
    pinnedTlsFetch,
    modelKey,
  });
}

async function runStreamingExample({
  apiKey,
  pinnedTlsFetch,
  modelKey,
}: CompletionExampleParams): Promise<Completion> {
  return sendCompletion({
    apiKey,
    stream: true,
    pinnedTlsFetch,
    modelKey,
  });
}

async function fetchAndVerifyGateway(
  client: AttestationClient,
): Promise<VerifiedGatewayAttestation> {
  // The Node client generates a fresh nonce and records the TLS peer's SPKI.
  // Fetching alone does not verify the returned evidence.
  const fetched = await client.fetchGatewayAttestation({
    signingAlgo: SIGNING_ALGO,
  });
  // Verify the quote, nonce, measurements, and signer/TLS binding. The deployment
  // callback checks image provenance after appCompose has been authenticated.
  const verified = await verifyGatewayAttestation({
    attestation: fetched.attestation,
    clientBinding: fetched.clientBinding,
    verifiers: {
      deployment: ({ appCompose }) =>
        verifyDeploymentImageProvenance({
          appCompose,
          imagePolicies: GATEWAY_IMAGE_POLICIES,
        }),
    },
  });
  console.log('Gateway deployment and image provenance: verified.');
  return verified;
}

async function fetchAndVerifyModelAttestations(
  client: AttestationClient,
): Promise<readonly VerifiedModelAttestation[]> {
  // These reports come through the Gateway, not a direct model TLS connection.
  // The fetch helper requests the signer-and-nonce layout without a TLS binding.
  const fetched = await client.fetchModelAttestations({
    model: MODEL,
    signingAlgo: SIGNING_ALGO,
  });
  if (fetched.attestations.length === 0) {
    throw new Error('Gateway returned no model attestations');
  }

  // All reports share the fetch helper's nonce. Verify each report's quote,
  // measurements, signing key, and GPU evidence when present before choosing a key.
  const verified = await Promise.all(
    fetched.attestations.map((attestation) =>
      verifyModelAttestation({
        attestation,
        clientBinding: fetched.clientBinding,
      }),
    ),
  );
  console.log(`Model deployments: verified ${verified.length}.`);
  return verified;
}

function findModelAttestationForEncryption(
  attestations: readonly VerifiedModelAttestation[],
): ModelAttestationWithPublicKey {
  // Choose an encryption recipient before Chat; there is no response signature
  // to select by yet. The E2EE helper also uses this key in the model-routing header.
  const attestation = attestations.find(
    (candidate): candidate is ModelAttestationWithPublicKey =>
      candidate.signer.signingAlgo === SIGNING_ALGO &&
      candidate.signingPublicKey !== undefined,
  );
  if (attestation === undefined) {
    throw new Error('No verified model public key is available for E2EE');
  }
  return attestation;
}

async function sendCompletion({
  apiKey,
  stream,
  pinnedTlsFetch,
  modelKey,
}: SendCompletionParams): Promise<Completion> {
  const request = new Request(new URL('chat/completions', BASE_URL), {
    method: 'POST',
    headers: {
      Authorization: `Bearer ${apiKey}`,
      'Content-Type': 'application/json',
      'Accept-Encoding': 'identity',
    },
    body: JSON.stringify({
      model: MODEL,
      messages: [{ role: 'user', content: 'Reply with the word ok.' }],
      stream,
      max_completion_tokens: 8,
    }),
  });
  // Encrypt protocol-supported fields to the verified model key and set the E2EE
  // headers. Other fields stay unchanged. This step does not send the request.
  // Each preparation creates a fresh response key pair, held by decryptResponse.
  const prepared = await prepareE2eeChatRequest({
    request,
    modelKey,
  });

  // The signature covers the encrypted request/response bodies, not the plaintext
  // shown to the user. Save the exact outgoing bytes without reserializing them.
  const requestBytes = await prepared.request.clone().arrayBuffer();
  const requestBody = new Uint8Array(requestBytes);
  const response = await pinnedTlsFetch(prepared.request);
  if (!response.ok) {
    throw new Error(
      `Chat request failed (${response.status}): ${await response.text()}`,
    );
  }
  // Keep an encrypted copy for verification; decrypt the other branch for display.
  // For SSE, consume both branches concurrently so output remains streaming.
  const encryptedResponse = response.clone();
  const decryptedResponse = await prepared.decryptResponse(response);
  const [responseBytes, plaintextBody] = await Promise.all([
    encryptedResponse.arrayBuffer(),
    readDecryptedCompletion({ response: decryptedResponse, stream }),
  ]);

  return {
    id: readCompletionId({ responseText: plaintextBody, stream }),
    label: stream ? 'Streaming completion' : 'Completion',
    requestBody,
    responseBody: new Uint8Array(responseBytes),
  };
}

async function readDecryptedCompletion({
  response,
  stream,
}: ReadDecryptedCompletionParams): Promise<string> {
  if (!stream) {
    const responseText = await response.text();
    console.log(responseText);
    return responseText;
  }
  if (response.body === null) {
    throw new Error('Streaming Chat response has no body');
  }
  const reader = response.body.pipeThrough(new TextDecoderStream()).getReader();
  let responseText = '';
  try {
    while (true) {
      const chunk = await reader.read();
      if (chunk.done) break;
      // Print raw decrypted SSE for this example; a UI would parse its data events.
      // Retain the text only to extract the completion ID, not for signature checks.
      process.stdout.write(chunk.value);
      responseText += chunk.value;
    }
  } finally {
    reader.releaseLock();
  }
  return responseText;
}

function createGatewayPinnedTlsFetch(
  tlsBinding: GatewayTlsBinding,
): typeof globalThis.fetch {
  if (tlsBinding.kind !== 'attested') {
    throw new Error('Expected TLS-bound Gateway evidence');
  }
  // Check each Chat connection against the attested SPKI. Reusing the original
  // attestation connection is unnecessary as long as the TLS key matches.
  return createPinnedTlsFetch(tlsBinding.spkiFingerprint);
}

async function verifyCompletionResponse({
  client,
  completion,
  gateway,
  model,
}: VerifyCompletionResponseParams): Promise<void> {
  // Retrieve the signature after the full response has been received. The ID is
  // only used for lookup; verification binds the saved bytes to an attested signer.
  const signature = await client.fetchCompletionSignature({
    completionId: completion.id,
    signingAlgo: SIGNING_ALGO,
  });

  if (signature.kind === 'provider_tee') {
    // Match the signature to the model selected for encryption, not another report.
    verifyModelResponse({
      requestBody: completion.requestBody,
      responseBody: completion.responseBody,
      signature,
      attestation: model,
    });
    console.log(`${completion.label}: verified model response signature.`);
  } else {
    // A Gateway signature authenticates the final bytes returned by the Gateway.
    // It does not prove model execution, even though model attestation passed above.
    verifyGatewayResponse({
      requestBody: completion.requestBody,
      responseBody: completion.responseBody,
      signature,
      attestation: gateway,
    });
    console.log(`${completion.label}: verified Gateway response signature.`);
  }
}

function readCompletionId({
  responseText,
  stream,
}: ReadCompletionIdParams): string {
  if (!stream) {
    return readJsonCompletionId(responseText);
  }

  for (const line of responseText.split(/\r\n|\n|\r/)) {
    if (!line.startsWith('data: ') || line === 'data: [DONE]') {
      continue;
    }
    try {
      return readJsonCompletionId(line.slice('data: '.length));
    } catch {
      // Later events may contain the completion ID.
    }
  }
  throw new Error('Streaming Chat response did not contain an id');
}

function readJsonCompletionId(text: string): string {
  const response: unknown = JSON.parse(text);
  if (
    typeof response !== 'object' ||
    response === null ||
    !('id' in response) ||
    typeof response.id !== 'string' ||
    response.id.length === 0
  ) {
    throw new Error('Chat response did not contain an id');
  }
  return response.id;
}

type Completion = {
  readonly id: string;
  readonly label: string;
  // Exact encrypted HTTP body bytes retained for response signature verification.
  readonly requestBody: Uint8Array;
  readonly responseBody: Uint8Array;
};

type CompletionExampleParams = {
  readonly apiKey: string;
  readonly pinnedTlsFetch: typeof globalThis.fetch;
  readonly modelKey: E2eeModelKey;
};

type SendCompletionParams = CompletionExampleParams & {
  readonly stream: boolean;
};

type ReadCompletionIdParams = {
  readonly responseText: string;
  readonly stream: boolean;
};

type ReadDecryptedCompletionParams = {
  readonly response: Response;
  readonly stream: boolean;
};

type VerifyCompletionResponseParams = {
  readonly client: AttestationClient;
  readonly completion: Completion;
  readonly gateway: VerifiedGatewayAttestation;
  readonly model: VerifiedModelAttestation;
};

type ModelAttestationWithPublicKey = VerifiedModelAttestation & {
  readonly signingPublicKey: string;
};
